server/src/main/java/keywhiz/service/resources/automation/v2/ExpirationExtractor.java

```diff
@@ -17,11 +21,57 @@
 import org.bouncycastle.util.io.pem.PemReader;
 
 import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Collections.list;
 
 /** Helper class to extract expirations from secrets contents (best effort) */
 public final class ExpirationExtractor {
   private ExpirationExtractor() {}
 
+  public static Instant expirationFromKeystore(String type, String password, byte[] content) {
+    KeyStore ks;
+    try {
+      ks = KeyStore.getInstance(type);
+    } catch (KeyStoreException e) {
+      // Should never occur (assuming JCE is installed)
+      throw Throwables.propagate(e);
+    }
+
+    try {
+      ks.load(new ByteArrayInputStream(content), password.toCharArray());
+    } catch (IOException | NoSuchAlgorithmException | CertificateException e) {
+      // Failed to parse
+      return null;
+    }
+
+    Instant earliest = null;
+    try {
+      for (String alias : list(ks.aliases())) {
+        Certificate[] chain = ks.getCertificateChain(alias);
+        if (chain == null) {
+          Certificate certificate = ks.getCertificate(alias);
+          if (certificate == null) {
+            // No certs in this entry
+            continue;
+          }
+          chain = new Certificate[]{certificate};
+        }
+        for (Certificate cert : chain) {
+          if (cert instanceof X509Certificate) {
+            X509Certificate c = (X509Certificate) cert;
+            if (earliest == null || c.getNotAfter().toInstant().isBefore(earliest)) {
+              earliest = c.getNotAfter().toInstant();
+            }
+          }
+        }
+      }
+    } catch (KeyStoreException e) {
+      // Should never occur (ks was initialized)
+      throw Throwables.propagate(e);
+    }
+
+    return earliest;
```

Comment:
Is there an instant where earliest is null? If so do we want to annotate the definition with @Nullable?